flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1826,4 +1838,30 @@ public SlotReport retrievePayload(ResourceID resourceID) {
 			return taskSlotTable.createSlotReport(getResourceID());
 		}
 	}
+
+	private static class TaskCompletionTracker {
+		private final Map<ExecutionAttemptID, Task> incompleteTasks;
+
+		private TaskCompletionTracker() {
+			incompleteTasks = new ConcurrentHashMap<>(8);
+		}
+
+		void trackTaskCompletion(Task task) {
+			incompleteTasks.put(task.getExecutionId(), task);
+			task.getTerminationFuture().thenRun(() -> incompleteTasks.remove(task.getExecutionId()));
```

Comment:
could we also rely on `TaskExecutor#unregisterTaskAndNotifyFinalState` to remove the task from the `TaskCompletionTracker`? That way we would not apply concurrent changes to `incompleteTasks`.

Reply:
This is also my concern. We already have `TaskExecutor#unregisterTaskAndNotifyFinalState` and `Task#getTerminationFuture`. I think we should unify them somehow.

Reply:
Andrey explained to me that we call `unregisterTaskAndNotifyFinalState` before the clean up of the `Task` has completed. I think this is a problem in the `Task's` lifecycle. Hence, one could fix this as a follow up and then simply use the `TerminationFuture` to signal the final state to the `TaskExecutor`. But this is out of scope for this PR.

Reply:
Yes, Andrey has explained to me before.
Never mind, it's just a mention when I saw this comment. I agree that we could follow up on it later.
BTW, we could also avoid concurrent modifications through `thenRunAsync` with `mainThreadExecutor` if we prefer relying on `task.getTerminationFuture()`.